### ISSUE TITLE
inference results

### ISSUE THREAD

**Issue by thepowerfuldeez** [keonlee9420/VAENAR-TTS#1]
Hi! Thank you for the port. Have you been able to get results on inference stage? I successfully train model, validation losses are decreasing, but at inference there's garbage.
I started logging log-probabilities of posterior and prior networks and see that they're also going down throughout training. Logpgobs in around -70000 for both networks which is very very small number, say zero in probability space.
Also if remove clipping kl divergence `torch.max(kl_divergence, torch.tensor(0., device=device))` something bad happens and kl goes negative, which is not possible in math point of view, but can be if our values are not valid distributions. Then I set n_samples to 4 and reduce batch_size to 8 but still get negative values. Pytorch's implementation of KLDivloss with log_targets=True always give 0 loss values....
so.. have you had any success?

**Comment by keonlee9420** [comment]
Hi @thepowerfuldeez , thank you for sharing the experiments. I guess you trained the model using the code which is still under implementation. Now I can make the model converge and may succeed on inferencing, too. I'll share the pre-trained models soon, so please stay tuned.

About the KL loss, you don't have to worry about it since I already checked and fixed the code. 
But for your reference, let me share my experiences. I also got negative values in some cases in other projects, and it can be seen as what you thought ([it is not actual distribution](https://discuss.pytorch.org/t/kl-divergence-produces-negative-values/16791/14)). But you can ignore it during the earlier steps in the training since it will follow the proper distribution after several steps. That's why the term `torch.max(kl_divergence, torch.tensor(0., device=device))` is added, and it's taking a role especially at the beginning of the training.

**Comment by thepowerfuldeez** [comment]
I also have my model converged according to validation losses and audio after hifigan. I took the state of your code from yesterday. I just wondering why my logprobs are so large
At inference length predictor is doing fine and also after decoding I can tell slightly the identity of the voice, but speech is completely gibberish, so the mel-spectrogram (same repeating pattern and noise).

**Comment by thepowerfuldeez** [comment]
I was trying to add ddp functionality but somehow attention is not converging in multi-gpu regime. Also mentioned experiments on 1 gpu after 300k iterations (800+ epochs)

**Comment by thepowerfuldeez** [comment]
and also what is the appropriate value for kl_loss? (using the code in repo) I got it increasing from 300 to 600 during the training (so when the r drops, loss increases), but mel and length losses are decreasing always.

**Comment by keonlee9420** [comment]
> I also have my model converged according to validation losses and audio after hifigan. I took the state of your code from yesterday. I just wondering why my logprobs are so large
> At inference length predictor is doing fine and also after decoding I can tell slightly the identity of the voice, but speech is completely gibberish, so the mel-spectrogram (same repeating pattern and noise).

In my understanding, it's because of the large KLD. Maybe the prior encoder doesn't work for the input latent, which is not a trained one (from posterior) for the model at inference time. Also, I found that the value of KLD is relatively mattered regarding the range and normalization of the input mel-spectrogram. What you referred may be using hparams for small input ranges, but the input was not normalized. But now, they are synced for the normalized inputs with modified mel-spectrogram calculations.

> I was trying to add ddp functionality but somehow attention is not converging in multi-gpu regime. Also mentioned experiments on 1 gpu after 300k iterations (800+ epochs)

I didn't check the multi-gpu conditions.

**Comment by thepowerfuldeez** [comment]
you're right, seems like with normalization kl converges faster. let's see where it gets
![image](https://user-images.githubusercontent.com/11796343/126195868-ba0b3b61-59ea-4797-9fdf-51f6fc8f3d0b.png)

**Comment by thepowerfuldeez** [comment]
it doesnt help. but what did help though is I rewrite prior network (there were some bugs in log_probability function – ordering of glow blocks and sign of log_det value). Even after first 5k iterations sampled attention looks diagonal! and speech is slightly distinguishable

**Comment by keonlee9420** [comment]
> you're right, seems like with normalization kl converges faster. let's see where it gets
> ![image](https://user-images.githubusercontent.com/11796343/126195868-ba0b3b61-59ea-4797-9fdf-51f6fc8f3d0b.png)

Right, but you should note that too strict kl from the beginning of the training may put the training too long or even prevent it at all. KL annealing needs to be scheduled carefully.
BTW, what is your logging interface? It looks fancy, and I want to try it too :)

> it doesnt help. but what did help though is I rewrite prior network (there were some bugs in log_probability function – ordering of glow blocks and sign of log_det value). Even after first 5k iterations sampled attention looks diagonal! and speech is slightly distinguishable

Nice! Could you elaborate more about the bugs that you found? Is it still relevant in the latest state of my code?

**Comment by thepowerfuldeez** [comment]
@keonlee9420 it is wandb.ai, been using it for 2 years and i'm in love with it! can log configs, check training from mobile phone, stop runs, it never lags as tb do etc. but in-house version costs money

yea, it is still relevant. to make prior network work you can check for changes in my fork here: https://github.com/thepowerfuldeez/VAENAR-TTS/blob/main/model/prior.py (and newly created glow.py as well). Glad it will help!

**Comment by keonlee9420** [comment]
I definitely should try. Thank you for sharing!
Did you succeed on inference, too? With the code from the fork.

**Comment by thepowerfuldeez** [comment]
@keonlee9420 yes but the quality is not perfect and length predictor is sometimes wrong by +- 50 frames which leads to bad audio endings. Also might be some issue with text processing at inference

**Comment by keonlee9420** [comment]
Sorry, I closed the issue by accident.. anyway I just apply the modified model and start training. Let me share the results soon. Of course you may share it first.

**Comment by keonlee9420** [comment]
@thepowerfuldeez I apply your update and find that it's working but have the same issues that you report. Maybe it's because the model does not yet be fully trained. Note that the preprocessing is different from yours and that makes the tiny variation in configs.

**Comment by thepowerfuldeez** [comment]
have you achieve better quality than that? https://voca.ro/17B6UKihLI90

**Comment by keonlee9420** [comment]
it seems similar to yours from the 355k model with batch size 32. did you train the hifi-gan, too?

**Comment by thepowerfuldeez** [comment]
nope, I have pretrained universal fre-gan model. you should also check it out, as it trains 5x faster than hifi-gan and have similar or better overall quality :)

**Comment by keonlee9420** [comment]
Close due to inactivity. You can reopen it anytime.

**Comment by shangeth** [comment]
After training, the Griffin-Lim gives better audio, but melgan included in this repo or melgan/hifi-gan from nemo gives garbage noise.

**Comment by keonlee9420** [comment]
@shangeth yeah right. that's why you should retrain the vocoder on the data preprocessed by the same way in training of the model.

**Comment by shangeth** [comment]
Hey @keonlee9420 , i understand the vocoder is needed to be trained together with the acoustic model. I don't understand why?. If the acoustic model generates Mel-Spectrograms, then any vocoder should be able to convert it to wav right?(even if its not of good quality audio, it should do a decent job). Is there any reason behind this?

**Comment by keonlee9420** [comment]
In fact, I also tried to match them, but my conclusion was that the method of [pre-emphasizing](https://github.com/keonlee9420/VAENAR-TTS/blob/3bfe693d1260b244c120f8eb11e3d33787c9aada/audio/audio.py#L146) and [normalizing](https://github.com/keonlee9420/VAENAR-TTS/blob/3bfe693d1260b244c120f8eb11e3d33787c9aada/audio/audio.py#L176) was a problem (in other words, it was not working to denormalize and inverse pre-emphasize the output of the model and then input to the pre-trained vocoder). After getting this, I just decided to stay on Griffin-Lim algorithm.

**Comment by wizardk** [comment]
> nope, I have pretrained universal fre-gan model. you should also check it out, as it trains 5x faster than hifi-gan and have similar or better overall quality :)

Hi @thepowerfuldeez , do you mean UnivNet?
